test/scripts/run_test.py

```diff
@@ -47,47 +49,58 @@ def analyse_instrumentation_test_result(result):
     return ok_num
 
 
-def analyse_unit_test_result(result):
-    total_num = 0
-    fail_num = 0
+def analyse_unit_test_result(result, mode, module):
+    xml_file_name = module + '-' + mode.value + '-result-' + TIMESTAMP + '.xml'
+    p = re.compile(".*?XML test result file generated at (.*?\.xml)")
+    dst_xml_path = os.path.join(os.path.split(result)[0], xml_file_name)
     with open(result, 'r') as f:
         for line in f:
-            m_failure = re.match('Tests run:\s+(\d+).*?Failures:\s+(\d+)', line)
-            m_succeed = re.match('OK.*?(\d+) test', line)
-            if m_failure is not None:
-                total_num = int(m_failure.group(1))
-                fail_num = int(m_failure.group(2))
-            if m_succeed is not None:
-                total_num = int(m_succeed.group(1))
-    return total_num, fail_num
-
-
-def exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file):
+            m = re.match(p, line)
+            if m is not None:
+                shutil.copyfile(m.group(1), dst_xml_path)
+    try:
+        xml_doc = minidom.parse(dst_xml_path)
+    except:
+        print "Error occured while reading test result."
+        return 0, 0
+    test_suite = xml_doc.documentElement
+    total_num = int(test_suite.attributes["tests"].value)
+    failed_num = int(test_suite.attributes["failures"].value)
+    error_num = int(test_suite.attributes["errors"].value)
+    success_num = total_num - failed_num - error_num
+    return total_num, success_num
+
+
+def exec_cmd(mode, cmd, device, target_package, result_file, logcat_file, cmd_path):
+    if device is not None:
+        os.environ['ANDROID_SERIAL'] = device
+    if mode == TestMode.UNIT:
+        subprocess.call([HOME_PATH + '/gradlew', '-q', 'clean'], cwd=cmd_path)
     adb = ['adb'] if device == None else ['adb', '-s', device]
     clean_logcat = ['logcat', '-c']
     subprocess.call(adb + clean_logcat)
     with open(result_file, 'a+') as rf:
-        subprocess.call(adb + am_cmd, stdout=rf)
+        subprocess.call(cmd, cwd=cmd_path, stdout=rf ,stderr=rf)
     logcat_cmd = ['logcat', '-d', target_package]
     with open(logcat_file, 'a+') as lf:
         subprocess.call(adb + logcat_cmd, stdout=lf)
```

Comment:
This function is named as `exec_cmd` which, i assume, means to execute a command basically. So running a logcat here doesn't seem clear and logical.

Reply:
done